README.txt
= ar_mailer

A two-phase delivery agent for ActionMailer

Rubyforge Project:

http://rubyforge.org/projects/seattlerb

Documentation:

http://seattlerb.org/ar_mailer

and for forked additions

http://github.com/adzap/ar_mailer/wikis

Bugs:

http://rubyforge.org/tracker/?func=add&group_id=1513&atid=5921

== About

Even delivering email to the local machine may take too long when you have to
send hundreds of messages.  ar_mailer allows you to store messages into the
database for later delivery by a separate process, ar_sendmail.

== Installing ar_mailer (forked)

Install the gem from GitHub gems server:

First, if you haven't already

  $ sudo gem sources -a http://gems.github.com

Then

  $ sudo gem install adzap-ar_mailer


== Usage

Go to your Rails project:

  $ cd your_rails_project

Create a new migration:

  $ ar_sendmail --create-migration

You'll need to redirect this into a file.  If you want a different name
provide the --table-name option.

Create a new model:

  $ ar_sendmail --create-model

You'll need to redirect this into a file.  If you want a different name
provide the --table-name option.

You'll need to be sure to set the From address for your emails.  Something
like:

  def list_send(recipient)
    from 'no_reply@example.com'
    # ...

Create an initializer or add this line to an existing one:

 require 'action_mailer/ar_mailer'

Edit config/environments/production.rb and set the delivery method:

  $ grep delivery_method config/environments/production.rb
  ActionMailer::Base.delivery_method = :activerecord

If you need to, you can set each mailer class delivery method individually:

  class MyMailer < ActionMailer::Base
    self.delivery_method = :activerecord
  end

This can be useful when using plugins like ExceptionNotification. Where it
might be foolish to tie the sending of the email alert to the database when the 
database might be causing the exception being raised. In this instance you could
override ExceptionNofitier delivery method to be smtp or set each the other 
mailer classes to use ARMailer.

Run ar_sendmail:

  $ ar_sendmail

You can also run it from cron with -o, or as a daemon with -d.

See <tt>ar_sendmail -h</tt> for full details.

=== Alternate Mail Storage

If you want to set the ActiveRecord model that emails will be stored in,
see ActionMailer::Base.email_class=


=== Help

See ar_sendmail -h for options to ar_sendmail.

NOTE: You may need to delete an smtp_tls.rb file if you have one lying
around.  ar_mailer supplies it own.

== Run as a service (init.d/rc.d scripts)

For Linux both script and demo config files are in share/linux. 
See ar_sendmail.conf for setting up your config. Copy the ar_sendmail file 
to /etc/init.d/ and make it executable. Then for Debian based distros run
'sudo update-rc.d ar_sendmail defaults' and it should work. Make sure you have 
the config file /etc/ar_sendmail.conf in place before starting.

For FreeBSD or NetBSD script is share/bsd/ar_sendmail. This is old and does not
support the config file unless someone wants to submit a patch.


History.txt
= 2.1.10

* Bugs fixed
  * Fix require 'thread' for newer version of Rubygems (groan)

= 2.1.9

* Merged in minimal environment load option (experimental)
* Add bundler support in linux init.d script

= 2.1.8

* Bugs fixed
  * Fix BCC addresses being visible to all recipients (laserlemon)

= 2.1.7

* Miscellaneous Updates
  * Added adzap-ar_mailer file to end the need for the config.gem :lib option

* Bugs fixed
  * Fix from address being set as address header object.

= 2.1.6

* Miscellaneous Updates
  * Use return-path as from address if set, as per default ActionMailer::Base SMTP delivery (Dustin Tinney)
  * Always wait full delay between deliveries (cyu)

* Bugs fixed
  * Non-zero return code returned in init.d script (hennk)
  * Use updated Timeout error class (hennk)

= 2.1.5

* Bugs fixed
  * Load ar_mailer after environment to fix issue with lazy loading of ActionMailer in Rails 2.3

= 2.1.4

* Bugs fixed
  * Explicitly require ar_mailer in ar_sendmail because its not getting loaded by the Rails environment for some reason

= 2.1.3

* Tests now pass on gem install
* Removed deprecated ActionMailer::ARMailer class
* Bugs fixed
  * Fixed issue with pre-loading ActionMailer. No use ActionMailer::Base.email_class directly rather than store in ARSendmail instance var so no need to pre-load ActionMailer.

= 2.1.2

* Bugs fixed
  * Require ar_mailer in ar_sendmail since the change to remove TableName and use email_class

= 2.1.1

 * Force gem rebuild

= 2.1.0

* Switched to using a Rails generator for migration and model files. The ar_sendmail options have been removed.

= 2.0.2

* Removed TableName option from ar_sendmail options as its redundant. The Rails environment gets loaded so the settings for email class also get loaded
* Bugs fixed
  * Email class reloading issue in development mode causing AR email class defaults to be lost when cached

= 2.0.1

* Added option to use smtp setting of :tls => false to disable TLS auto start in Ruby 1.8.7+
* Removed some cruft which can be handled by ActiveSupport

= 2.0.0

* Removed need to use ARMailer subclass. Just set the delivery method and you are ready to go. Backwards compatible with a deprecation notice if you subclass old ARMailer class.
* Only include SMTP TLS patch if Ruby version < 1.8.7 as it has an alternative. Changes based on Calvin Yu's [cyu] fork.
* Renamed default migration name to the modern Rails default
* Only authenticate if emails waiting to be sent
* Added --version switch to ar_sendmail binary
* Created a lighthouse account for this project (adzap fork only). See README.

= 1.4.4

* Exit init.d script with message if no mailers defined.

= 1.4.3

* Bugs fixed
  * Replaced mistaken call to log when removing pid file artifact for
    non-running daemon

= 1.4.2

* New Features
  * Added Ruby based linux init.d script for handling daemon startup using yaml
    config file. See files share/linux/ar_sendmail and ar_sendmail.conf
* Bugs fixed
  * Proper handling for relative and absolute paths for the pid file
* Removed hoe dependency since we need the explicit gemspec file for github and
  not deploying to RubyForge its not as useful.
* Moved old BSD rc.d script to share/bsd folder
* Updated README with github gem install, docs and init script info

= 1.4.1

* Bugs fixed
  * Daemon failed on startup fixed with expanding full path of pid file

= 1.4.0

* Forked gem and published on GitHub (gem sources -a http://gems.github.com)
* New Features
  * Added pid file creation on daemonize with command line option to specify pid filename [Dylan Egan]

= 1.3.1

* Fix bug #12530, gmail causes SSL errors.  Submitted by Kyle Maxwell
  and Alex Ostleitner.
* Try ActionMailer::Base::server_settings then ::smtp_settings.  Fixes
  bug #12516.  Submitted by Alex Ostleitner.

= 1.3.0

* New Features
  * Added automatic mail queue cleanup.
    * MAY CAUSE LOSS OF DATA.  If you haven't run ar_sendmail within
      the expiry time, set it to 0.
* Bugs fixed
  * Authentication errors are now handled by retrying once.

= 1.2.0

* Bugs fixed
  * Handle SMTPServerBusy by backing off @delay seconds then re-queueing
  * Allow email delivery class to be set in ARMailer.
  * ar_sendmail --mailq works with --table-name now.
* Miscellaneous Updates
  * Added documentation to require 'action_mailer/ar_mailer' in
    instructions.
  * Moved to ZSS p4 repository
  * Supports TLS now.  Requested by Dave Thomas. smtp_tls.rb from Kyle
    Maxwell & etc.

= 1.1.0

* Features
  * Added --chdir to set rails directory
  * Added --environment to set RAILS_ENV
  * Exits cleanly on TERM or INT signals
  * Added FreeBSD rc.d script
  * Exceptions during SMTP sending are now logged
  * No longer waits if sending email took too long
* Bugs fixed
  * Fixed last send attempt in --mailq
  * Better SMTP error handling
    * Messages are removed from the queue on 5xx errors
    * Added Net::SMTP.reset to avoid needing to recreate the connection

= 1.0.1

* Bugs fixed
  * From and to of email destination were swapped

= 1.0.0

* Birthday


test/test_helper.rb
require 'test/unit'
require 'rubygems'
require 'active_support'
require 'test/resources/action_mailer'
require 'minitest/autorun'
require 'mocha'

require 'action_mailer/ar_mailer'
require 'action_mailer/ar_sendmail'


test/action_mailer.rb
require 'net/smtp'
require 'smtp_tls' unless Net::SMTP.instance_methods.include?("enable_starttls_auto")
require 'time'

class Net::SMTP

  @reset_called = 0

  @deliveries = []

  @send_message_block = nil

  @start_block = nil

  class << self

    attr_reader :deliveries
    attr_reader :send_message_block
    attr_accessor :reset_called

    send :remove_method, :start

  end

  def self.on_send_message(&block)
    @send_message_block = block
  end

  def self.on_start(&block)
    if block_given?
      @start_block = block
    else
      @start_block
    end
  end

  def self.clear_on_start
    @start_block = nil
  end

  def self.reset
    deliveries.clear
    on_start
    on_send_message
    @reset_called = 0
  end

  def start(*args)
    self.class.on_start.call if self.class.on_start
    yield self
  end

  alias test_old_reset reset if instance_methods.include? 'reset'

  def reset
    self.class.reset_called += 1
  end

  alias test_old_send_message send_message

  def send_message(mail, to, from)
    return self.class.send_message_block.call(mail, to, from) unless
      self.class.send_message_block.nil?
    self.class.deliveries << [mail, to, from]
    return "queued"
  end

end

##
# Stub for ActionMailer::Base

module ActionMailer; end

class ActionMailer::Base

  @server_settings = {}

  class << self
    cattr_accessor :email_class
    attr_accessor :delivery_method
  end

  def self.logger
    o = Object.new
    def o.info(arg) end
    return o
  end

  def self.method_missing(meth, *args)
    meth.to_s =~ /deliver_(.*)/
    super unless $1
    new($1, *args).deliver!
  end

  def self.reset
    server_settings.clear
    self.email_class = Email
  end

  def self.server_settings
    @server_settings
  end

  def initialize(meth = nil)
    send meth if meth
  end

  def deliver!
    perform_delivery_activerecord @mail
  end

end

##
# Stub for an ActiveRecord model

class Email

  START = Time.parse 'Thu Aug 10 2006 11:19:48'

  attr_accessor :from, :to, :mail, :last_send_attempt, :created_on, :id

  @records = []
  @id = 0

  class << self; attr_accessor :records, :id; end

  def self.create(record)
    record = new record[:from], record[:to], record[:mail],
                 record[:last_send_attempt]
    records << record
    return record
  end

  def self.destroy_all(conditions)
    timeout = conditions.last
    found = []

    records.each do |record|
      next if record.last_send_attempt == 0
      next if record.created_on == 0
      next unless record.created_on < timeout
      record.destroy
      found << record
    end

    found
  end

  def self.find(_, conditions = nil)
    return records if conditions.nil?
    now = Time.now.to_i - 300
    return records.select do |r|
      r.last_send_attempt < now
    end
  end

  def self.reset
    @id = 0
    records.clear
  end

  def initialize(from, to, mail, last_send_attempt = nil)
    @from = from
    @to = to
    @mail = mail
    @id = self.class.id += 1
    @created_on = START + @id
    @last_send_attempt = last_send_attempt || 0
  end

  def destroy
    self.class.records.delete self
    self.freeze
  end

  def ==(other)
    other.id == id
  end

  def save
  end

end

Mail = Email

class Newsletter < Email; end

class String
  def classify
    self
  end

  def tableize
    self.downcase
  end

end



test/test_armailer.rb
require File.expand_path(File.dirname(__FILE__) + '/test_helper')

class Mailer < ActionMailer::Base
  self.delivery_method = :activerecord

  def mail
    @mail = Object.new
    def @mail.encoded() 'email' end
    def @mail.from() ['nobody@example.com'] end
    def @mail.[](key) {'return-path' => $return_path, 'from' => 'nobody@example.com'}[key] end
    def @mail.destinations() %w[user1@example.com user2@example.com] end
    def @mail.ready_to_send() end
  end

end

class TestARMailer < Test::Unit::TestCase

  def setup
    $return_path = nil
    Mailer.email_class = Email

    Email.records.clear
    Newsletter.records.clear
  end

  def test_self_email_class_equals
    Mailer.email_class = Newsletter

    Mailer.deliver_mail

    assert_equal 2, Newsletter.records.length
  end

  def test_perform_delivery_activerecord_when_return_path_is_present
    $return_path = stub(:spec => 'return-path@example.com')
    Mailer.deliver_mail

    assert_equal 2, Email.records.length
    record = Email.records.first
    assert_equal 'return-path@example.com', record.from
  end

  def test_perform_delivery_activerecord
    Mailer.deliver_mail

    assert_equal 2, Email.records.length

    record = Email.records.first
    assert_equal 'email', record.mail
    assert_equal 'user1@example.com', record.to
    assert_equal 'nobody@example.com', record.from

    assert_equal 'user2@example.com', Email.records.last.to
  end

end



lib/smtp_tls.rb
# Original code believed public domain from ruby-talk or ruby-core email.
# Modifications by Kyle Maxwell <kyle@kylemaxwell.com> used under MIT license.

require "openssl"
require "net/smtp"

# :stopdoc:

class Net::SMTP

  class << self
    send :remove_method, :start
  end

  def self.start( address, port = nil,
                  helo = 'localhost.localdomain',
                  user = nil, secret = nil, authtype = nil, use_tls = false,
                  &block) # :yield: smtp
    new(address, port).start(helo, user, secret, authtype, use_tls, &block)
  end

  alias tls_old_start start

  def start( helo = 'localhost.localdomain',
             user = nil, secret = nil, authtype = nil, use_tls = false ) # :yield: smtp
    start_method = use_tls ? :do_tls_start : :do_start
    if block_given?
      begin
        send start_method, helo, user, secret, authtype
        return yield(self)
      ensure
        do_finish
      end
    else
      send start_method, helo, user, secret, authtype
      return self
    end
  end

  private

  def do_tls_start(helodomain, user, secret, authtype)
    raise IOError, 'SMTP session already started' if @started
    check_auth_args user, secret, authtype if user or secret

    sock = timeout(@open_timeout) { TCPSocket.open(@address, @port) }
    @socket = Net::InternetMessageIO.new(sock)
    @socket.read_timeout = 60 #@read_timeout
    @socket.debug_output = STDERR #@debug_output

    check_response(critical { recv_response() })
    do_helo(helodomain)

    raise 'openssl library not installed' unless defined?(OpenSSL)
    starttls
    ssl = OpenSSL::SSL::SSLSocket.new(sock)
    ssl.sync_close = true
    ssl.connect
    @socket = Net::InternetMessageIO.new(ssl)
    @socket.read_timeout = 60 #@read_timeout
    @socket.debug_output = STDERR #@debug_output
    do_helo(helodomain)

    authenticate user, secret, authtype if user
    @started = true
  ensure
    unless @started
      # authentication failed, cancel connection.
        @socket.close if not @started and @socket and not @socket.closed?
      @socket = nil
    end
  end

  def do_helo(helodomain)
     begin
      if @esmtp
        ehlo helodomain
      else
        helo helodomain
      end
    rescue Net::ProtocolError
      if @esmtp
        @esmtp = false
        @error_occured = false
        retry
      end
      raise
    end
  end

  def starttls
    getok('STARTTLS')
  end

  alias tls_old_quit quit

  def quit
    begin
      getok('QUIT')
    rescue EOFError
    end
  end

end unless Net::SMTP.private_method_defined? :do_tls_start or
           Net::SMTP.method_defined? :tls?


lib/action_mailer/ar_mailer.rb
##
# Adds sending email through an ActiveRecord table as a delivery method for
# ActionMailer.
#

class ActionMailer::Base

  ##
  # Set the email class for deliveries. Handle class reloading issues which prevents caching the email class.
  #
  @@email_class_name = 'Email'

  def self.email_class=(klass)
    @@email_class_name = klass.to_s
  end

  def self.email_class
    @@email_class_name.constantize
  end

  ##
  # Adds +mail+ to the Email table.  Only the first From address for +mail+ is
  # used.

  def perform_delivery_activerecord(mail)
    destinations = mail.destinations
    mail.ready_to_send
    sender = (mail['return-path'] && mail['return-path'].spec) || mail.from.first
    destinations.each do |destination|
      self.class.email_class.create :mail => mail.encoded, :to => destination, :from => sender
    end
  end

end


lib/action_mailer/ar_sendmail.rb
require 'optparse'
require 'net/smtp'
require 'smtp_tls' unless Net::SMTP.instance_methods.include?("enable_starttls_auto")

##
# Hack in RSET

module Net # :nodoc:
class SMTP # :nodoc:

  unless instance_methods.include? 'reset' then
    ##
    # Resets the SMTP connection.

    def reset
      getok 'RSET'
    end
  end

end
end

##
# ActionMailer::ARSendmail delivers email from the email table to the
# SMTP server configured in your application's config/environment.rb.
# ar_sendmail does not work with sendmail delivery.
#
# ar_mailer can deliver to SMTP with TLS using smtp_tls.rb borrowed from Kyle
# Maxwell's action_mailer_optional_tls plugin.  Simply set the :tls option in
# ActionMailer::Base's smtp_settings to true to enable TLS.
#
# See ar_sendmail -h for the full list of supported options.
#
# The interesting options are:
# * --daemon
# * --mailq

module ActionMailer; end

class ActionMailer::ARSendmail

  class RailsEnvironmentFailed < StandardError; end
  class MinimalEnvironmentFailed < StandardError; end

  ##
  # The version of ActionMailer::ARSendmail you are running.

  VERSION = '2.1.10'

  ##
  # Maximum number of times authentication will be consecutively retried

  MAX_AUTH_FAILURES = 2

  ##
  # Email delivery attempts per run

  attr_accessor :batch_size

  ##
  # Seconds to delay between runs

  attr_accessor :delay

  ##
  # Maximum age of emails in seconds before they are removed from the queue.

  attr_accessor :max_age

  ##
  # Be verbose

  attr_accessor :verbose


  ##
  # True if only one delivery attempt will be made per call to run

  attr_reader :once

  ##
  # Times authentication has failed

  attr_accessor :failed_auth_count

  @@pid_file = nil

  def self.remove_pid_file
    if @@pid_file
      require 'shell'
      sh = Shell.new
      sh.rm @@pid_file
    end
  end

  ##
  # Prints a list of unsent emails and the last delivery attempt, if any.
  #
  # If ActiveRecord::Timestamp is not being used the arrival time will not be
  # known.  See http://api.rubyonrails.org/classes/ActiveRecord/Timestamp.html
  # to learn how to enable ActiveRecord::Timestamp.

  def self.mailq
    emails = ActionMailer::Base.email_class.find :all

    if emails.empty? then
      puts "Mail queue is empty"
      return
    end

    total_size = 0

    puts "-Queue ID- --Size-- ----Arrival Time---- -Sender/Recipient-------"
    emails.each do |email|
      size = email.mail.length
      total_size += size

      create_timestamp = email.created_on rescue
                         email.created_at rescue
                         Time.at(email.created_date) rescue # for Robot Co-op
                         nil

      created = if create_timestamp.nil? then
                  '             Unknown'
                else
                  create_timestamp.strftime '%a %b %d %H:%M:%S'
                end

      puts "%10d %8d %s  %s" % [email.id, size, created, email.from]
      if email.last_send_attempt > 0 then
        puts "Last send attempt: #{Time.at email.last_send_attempt}"
      end
      puts "                                         #{email.to}"
      puts
    end

    puts "-- #{total_size/1024} Kbytes in #{emails.length} Requests."
  end

  ##
  # Processes command line options in +args+

  def self.process_args(args)
    name = File.basename $0

    options = {}
    options[:Chdir] = '.'
    options[:Daemon] = false
    options[:Delay] = 60
    options[:MaxAge] = 86400 * 7
    options[:Once] = false
    options[:RailsEnv] = ENV['RAILS_ENV']
    options[:Pidfile] = options[:Chdir] + '/log/ar_sendmail.pid'
    options[:Minimal] = false

    opts = OptionParser.new do |opts|
      opts.banner = "Usage: #{name} [options]"
      opts.separator ''

      opts.separator "#{name} scans the email table for new messages and sends them to the"
      opts.separator "website's configured SMTP host."
      opts.separator ''
      opts.separator "#{name} must be run from a Rails application's root."

      opts.separator ''
      opts.separator 'Sendmail options:'

      opts.on("-b", "--batch-size BATCH_SIZE",
              "Maximum number of emails to send per delay",
              "Default: Deliver all available emails", Integer) do |batch_size|
        options[:BatchSize] = batch_size
      end

      opts.on(      "--delay DELAY",
              "Delay between checks for new mail",
              "in the database",
              "Default: #{options[:Delay]}", Integer) do |delay|
        options[:Delay] = delay
      end

      opts.on(      "--max-age MAX_AGE",
              "Maxmimum age for an email. After this",
              "it will be removed from the queue.",
              "Set to 0 to disable queue cleanup.",
              "Default: #{options[:MaxAge]} seconds", Integer) do |max_age|
        options[:MaxAge] = max_age
      end

      opts.on("-o", "--once",
              "Only check for new mail and deliver once",
              "Default: #{options[:Once]}") do |once|
        options[:Once] = once
      end

      opts.on("-d", "--daemonize",
              "Run as a daemon process",
              "Default: #{options[:Daemon]}") do |daemon|
        options[:Daemon] = true
      end

      opts.on(      "--minimal",
              "Load a minimal environment with settings from config/email.yml",
              "Default: #{options[:Minimal]}") do |minimal|
        options[:Minimal] = true
      end

      opts.on("-p", "--pidfile PIDFILE",
              "Set the pidfile location",
              "Default: #{options[:Chdir]}#{options[:Pidfile]}", String) do |pidfile|
        options[:Pidfile] = pidfile
      end

      opts.on(      "--mailq",
              "Display a list of emails waiting to be sent") do |mailq|
        options[:MailQ] = true
      end

      opts.separator ''
      opts.separator 'Setup Options:'

      opts.separator ''
      opts.separator 'Generic Options:'

      opts.on("-c", "--chdir PATH",
              "Use PATH for the application path",
              "Default: #{options[:Chdir]}") do |path|
        usage opts, "#{path} is not a directory" unless File.directory? path
        usage opts, "#{path} is not readable" unless File.readable? path
        options[:Chdir] = path
      end

      opts.on("-e", "--environment RAILS_ENV",
              "Set the RAILS_ENV constant",
              "Default: #{options[:RailsEnv]}") do |env|
        options[:RailsEnv] = env
      end

      opts.on("-v", "--[no-]verbose",
              "Be verbose",
              "Default: #{options[:Verbose]}") do |verbose|
        options[:Verbose] = verbose
      end

      opts.on("-h", "--help",
              "You're looking at it") do
        usage opts
      end

      opts.on("--version", "Version of ARMailer") do
        usage "ar_mailer #{VERSION} (adzap fork)"
      end

      opts.separator ''
    end

    opts.parse! args

    ENV['RAILS_ENV'] = options[:RailsEnv]

    begin
      options[:Minimal] ? load_minimal_environment(options[:Chdir]) : load_rails_environment(options[:Chdir])
    rescue RailsEnvironmentFailed
      usage opts, <<-EOF
#{name} must be run from a Rails application's root to deliver email.
#{Dir.pwd} does not appear to be a Rails application root.
      EOF
    rescue MinimalEnvironmentFailed => e
      usage opts, "Minimal environment loading has failed with error '#{e.message}'. Check minimal environment instructions or use the normal Rails environment loading."
    end

    return options
  end

  # Load full Rails environment
  #
  def self.load_rails_environment(base_path)
    Dir.chdir(base_path) do
      require 'config/environment'
      require 'action_mailer/ar_mailer'
    end
  rescue LoadError
    raise RailsEnvironmentFailed
  end

  # Load a minimal environment to save memory by not loading the entire Rails app.
  # Requires a vendored Rails or bundler, and a mailer config at config/email.yml.
  #
  def self.load_minimal_environment(base_path)
    Dir.chdir(base_path) do
      Dir.glob('vendor/rails/*/lib').each { |dir| $:.unshift File.expand_path(dir) }
      require 'thread'
      require 'yaml'
      require 'erb'
      require 'active_record'
      require 'action_mailer'
      require 'action_mailer/ar_mailer'
      require 'app/models/email'

      env = ENV['RAILS_ENV']

      logger = ActiveSupport::BufferedLogger.new(File.join('log', "#{env}.log"))
      logger.level = ActiveSupport::BufferedLogger.const_get((env == 'production' ? 'info' : 'debug').upcase)
      ActiveRecord::Base.logger = ActionMailer::Base.logger = logger

      db_config = read_config('config/database.yml')
      ActiveRecord::Base.establish_connection db_config[env]

      mailer_config = read_config('config/email.yml')
      ActionMailer::Base.smtp_settings = mailer_config[env].symbolize_keys
    end
  rescue => e
    raise MinimalEnvironmentFailed, e.message
  end
      def default_log_path
        File.join(root_path, 'log', "#{environment}.log")
      end

      def default_log_level
        environment == 'production' ? :info : :debug
      end

  # Open yaml config file and parse with ERB
  #
  def self.read_config(file)
    YAML::load(ERB.new(IO.read(file)).result)
  end

  ##
  # Processes +args+ and runs as appropriate

  def self.run(args = ARGV)
    options = process_args args

    if options.include? :MailQ then
      mailq
      exit
    end

    if options[:Daemon] then
      require 'webrick/server'
      @@pid_file = File.expand_path(options[:Pidfile], options[:Chdir])
      if File.exists? @@pid_file
        # check to see if process is actually running
        pid = ''
        File.open(@@pid_file, 'r') {|f| pid = f.read.chomp }
        if system("ps -p #{pid} | grep #{pid}") # returns true if process is running, o.w. false
          $stderr.puts "Warning: The pid file #{@@pid_file} exists and ar_sendmail is running. Shutting down."
          exit -1
        else
          # not running, so remove existing pid file and continue
          self.remove_pid_file
          $stderr.puts "ar_sendmail is not running. Removing existing pid file and starting up..."
        end
      end
      WEBrick::Daemon.start
      File.open(@@pid_file, 'w') {|f| f.write("#{Process.pid}\n")}
    end

    new(options).run

  rescue SystemExit
    raise
  rescue SignalException
    exit
  rescue Exception => e
    $stderr.puts "Unhandled exception #{e.message}(#{e.class}):"
    $stderr.puts "\t#{e.backtrace.join "\n\t"}"
    exit -2
  end

  ##
  # Prints a usage message to $stderr using +opts+ and exits

  def self.usage(opts, message = nil)
    if message then
      $stderr.puts message
      $stderr.puts
    end

    $stderr.puts opts
    exit 1
  end

  ##
  # Creates a new ARSendmail.
  #
  # Valid options are:
  # <tt>:BatchSize</tt>:: Maximum number of emails to send per delay
  # <tt>:Delay</tt>:: Delay between deliver attempts
  # <tt>:Once</tt>:: Only attempt to deliver emails once when run is called
  # <tt>:Verbose</tt>:: Be verbose.

  def initialize(options = {})
    options[:Delay] ||= 60
    options[:MaxAge] ||= 86400 * 7

    @batch_size = options[:BatchSize]
    @delay = options[:Delay]
    @once = options[:Once]
    @verbose = options[:Verbose]
    @max_age = options[:MaxAge]

    @failed_auth_count = 0
  end

  ##
  # Removes emails that have lived in the queue for too long.  If max_age is
  # set to 0, no emails will be removed.

  def cleanup
    return if @max_age == 0
    timeout = Time.now - @max_age
    conditions = ['last_send_attempt > 0 and created_on < ?', timeout]
    mail = ActionMailer::Base.email_class.destroy_all conditions

    log "expired #{mail.length} emails from the queue"
  end

  ##
  # Delivers +emails+ to ActionMailer's SMTP server and destroys them.

  def deliver(emails)
    settings = [
      smtp_settings[:domain],
      (smtp_settings[:user] || smtp_settings[:user_name]),
      smtp_settings[:password],
      smtp_settings[:authentication]
    ]

    smtp = Net::SMTP.new(smtp_settings[:address], smtp_settings[:port])
    if smtp.respond_to?(:enable_starttls_auto)
      smtp.enable_starttls_auto unless smtp_settings[:tls] == false
    else
      settings << smtp_settings[:tls]
    end

    smtp.start(*settings) do |session|
      @failed_auth_count = 0
      until emails.empty? do
        email = emails.shift
        begin
          res = session.send_message email.mail, email.from, email.to
          email.destroy
          log "sent email %011d from %s to %s: %p" %
                [email.id, email.from, email.to, res]
        rescue Net::SMTPFatalError => e
          log "5xx error sending email %d, removing from queue: %p(%s):\n\t%s" %
                [email.id, e.message, e.class, e.backtrace.join("\n\t")]
          email.destroy
          session.reset
        rescue Net::SMTPServerBusy => e
          log "server too busy, stopping delivery cycle"
          return
        rescue Net::SMTPUnknownError, Net::SMTPSyntaxError, TimeoutError, Timeout::Error => e
          email.last_send_attempt = Time.now.to_i
          email.save rescue nil
          log "error sending email %d: %p(%s):\n\t%s" %
                [email.id, e.message, e.class, e.backtrace.join("\n\t")]
          session.reset
        end
      end
    end
  rescue Net::SMTPAuthenticationError => e
    @failed_auth_count += 1
    if @failed_auth_count >= MAX_AUTH_FAILURES then
      log "authentication error, giving up: #{e.message}"
      raise e
    else
      log "authentication error, retrying: #{e.message}"
    end
    sleep delay
  rescue Net::SMTPServerBusy, SystemCallError, OpenSSL::SSL::SSLError
    # ignore SMTPServerBusy/EPIPE/ECONNRESET from Net::SMTP.start's ensure
  end

  ##
  # Prepares ar_sendmail for exiting

  def do_exit
    log "caught signal, shutting down"
    self.class.remove_pid_file
    exit 130
  end

  ##
  # Returns emails in email_class that haven't had a delivery attempt in the
  # last 300 seconds.

  def find_emails
    options = { :conditions => ['last_send_attempt < ?', Time.now.to_i - 300] }
    options[:limit] = batch_size unless batch_size.nil?
    mail = ActionMailer::Base.email_class.find :all, options

    log "found #{mail.length} emails to send"
    mail
  end

  ##
  # Installs signal handlers to gracefully exit.

  def install_signal_handlers
    trap 'TERM' do do_exit end
    trap 'INT'  do do_exit end
  end

  ##
  # Logs +message+ if verbose

  def log(message)
    $stderr.puts message if @verbose
    ActionMailer::Base.logger.info "ar_sendmail: #{message}"
  end

  ##
  # Scans for emails and delivers them every delay seconds.  Only returns if
  # once is true.

  def run
    install_signal_handlers

    loop do
      begin
        cleanup
        emails = find_emails
        deliver(emails) unless emails.empty?
      rescue ActiveRecord::Transactions::TransactionError
      end
      break if @once
      sleep @delay
    end
  end

  ##
  # Proxy to ActionMailer::Base::smtp_settings.  See
  # http://api.rubyonrails.org/classes/ActionMailer/Base.html
  # for instructions on how to configure ActionMailer's SMTP server.
  #
  # Falls back to ::server_settings if ::smtp_settings doesn't exist for
  # backwards compatibility.

  def smtp_settings
    ActionMailer::Base.smtp_settings rescue ActionMailer::Base.server_settings
  end

end
